### ISSUE TITLE
KE-T5를 이용해 키워드 추출 연구를 진행하려는데 궁금한 점이 있습니다.

### ISSUE THREAD

**Issue by ahj6377** [AIRC-KETI/ke-t5#10]
Summarization 관련 부분 코드 

---------------------------------------------------------------

infer = loaded.signatures["serving_default"]
input_str = """“입력문서"""

input_str_topic = "summarize_topic: " + input_str
input_str_summary = "summarize_summary: " + input_str

x = tf.constant([input_str_topic])

result = infer(x)
print([out.decode('utf-8') for out in result['inputs'].numpy()])
print([out.decode('utf-8') for out in result['outputs'].numpy()])

---------------------------------------------------------------

이 부분에서 infer 자체가 입력 text를 받는곳이란걸로 이해했는데,

텐서플로우 모델로

![dataframe](https://user-images.githubusercontent.com/24570705/235860320-270155de-319d-45c9-8c33-bd183a89c22b.png)

이러한 형태의 데이터를 학습시키려고 train_x, train_y를 불러서 정의하고

-------------------------------------------------------------------------------
model = infer

optimizer = tf.keras.optimizers.Adam(learning_rate = 0.00001)
loss = tf.keras.losses.MeanSquaredError()

model.compile(optimizer = optimizer, loss = loss, metrics = ["accuracy"])
-------------------------------------------------------------------------------

일단 시도해보려고 위같이 model compile을 시도했는데 
'WrappedFunction' object has no attribute 'compile' 에러가 나왔습니다.
혹시 load된 파일이 체크포인트 자체가 아닌가요? 텐서플로우 모델로 직접 fine tuning을 하려면 어느파일을 불러 모델을 정의해야하는지 알 수 있을까요?
일단 모델은 t5.1.1.base.gin.ke.ke_v100_span_corruption.ke_t5_nikl_summary_mixture_equal 모델을 불러왔습니다.

**Comment by kimsan0622** [comment]
넵. 사용하신 파일은 모델 체크포인트가 아니라 사용하기 편하게하기 위하여 SavedModel 포맷(tokenization부터 decoding까지 그래프가 다 연결된 상태)으로 저장된 모델입니다.
해당 포맷에 대한 자세한 설명은 [Tensorflow SavedModel](https://www.tensorflow.org/guide/saved_model?hl=ko)을 참고하시길 바랍니다.

만약 [Mesh Tensorflow](https://github.com/tensorflow/mesh)(model parallel을 위한 tensorflow library)에 익숙하시다면 해당 모델의 [checkpoint](https://github.com/AIRC-KETI/ke-t5#pretrained-models)의 `Download URL(Tensorflow)` 칼럼을 다운받아 이용하시면 됩니다.

Mesh Tensorflow에 익숙하지 않아 Keras나 Pytorch를 사용하신다면 [같은 링크](https://github.com/AIRC-KETI/ke-t5#pretrained-models)의 `Huggingface Transformers` 칼럼에 써져있는 값을 pretrained 모델으로 사용하시면 됩니다.

간단한 tokenizer와 model load 예시를 들어보겠습니다. 자세한 내용은 [Huggingface NLP Course](https://huggingface.co/learn/nlp-course/chapter3/3?fw=tf)를 참고하세요.

먼저 transformers, sentencepiece, protobuf를 설치해줍니다. (예전에는 protobuf와 sentencepiece간 버전 문제가 좀 있었는데, 요즘엔 해결된거 같습니다. 만약 에러가 난다면 protobuf<=3.20.0로 설치해주세요.)
```bash
pip install transformers sentencepiece protobuf
```

pytorch의 경우
```python
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

model_name_or_path = "KETI-AIR/ke-t5-base"
tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path)
```

tensorflow keras의 경우
```python
from transformers import AutoTokenizer, TFAutoModelForSeq2SeqLM

model_name_or_path = "KETI-AIR/ke-t5-base"
tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
model = TFAutoModelForSeq2SeqLM.from_pretrained(model_name_or_path)
```

하시려는 task를 보니 token level classification이 필요해 보입니다. BIO2 태깅을 한 다음, token level classification을 하시거나, head에 CRF를 붙여 NER처럼 진행하시면 될 것 같습니다.
token level classification은 transformers github에 예시가 잘 되어 있습니다.
[Tensorflow Token Classification](https://github.com/huggingface/transformers/tree/main/examples/tensorflow/token-classification)
[Pytorch Token Classification](https://github.com/huggingface/transformers/tree/main/examples/pytorch/token-classification)

T5는 Encoder, Decoder 모델이기 때문에, Token Classification을 위해서는 Encoder에 classification head를 붙여줘야 합니다. 이에 대한 예시는 Pytorch로만 들도록 하겠습니다.

Huggingface ~~~ForTokenClassification과 같이 linear layer 한개만 classification head로 사용할 경우
```python
from transformers import T5EncoderModel
from transformers.modeling_outputs import TokenClassifierOutput

import torch
import torch.nn as nn
from torch.nn import CrossEntropyLoss


class T5EncoderForTokenClassification(T5EncoderModel):
    def init(self, config):
        super().init(config)
        self.num_labels = config.num_labels

        self.dropout = nn.Dropout(config.dropout_rate)
        self.classifier = nn.Linear(config.d_model, config.num_labels)

    def forward(
        self,
        input_ids=None,
        attention_mask=None,
        head_mask=None,
        inputs_embeds=None,
        labels=None,
        output_attentions=None,
        output_hidden_states=None,
        return_dict=None,
    ):
        return_dict = return_dict if return_dict is not None else self.config.use_return_dict

        outputs = self.encoder(
            input_ids=input_ids,
            attention_mask=attention_mask,
            inputs_embeds=inputs_embeds,
            head_mask=head_mask,
            output_attentions=output_attentions,
            output_hidden_states=output_hidden_states,
            return_dict=return_dict,
        )

        sequence_output = outputs[0]

        sequence_output = self.dropout(sequence_output)
        logits = self.classifier(sequence_output)

        loss = None
        if labels is not None:
            loss_fct = CrossEntropyLoss()
            # Only keep active parts of the loss
            if attention_mask is not None:
                active_loss = attention_mask.view(-1) == 1
                active_logits = logits.view(-1, self.num_labels)
                active_labels = torch.where(
                    active_loss, labels.view(-1), torch.tensor(loss_fct.ignore_index).type_as(labels)
                )
                loss = loss_fct(active_logits, active_labels)
            else:
                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))

        if not return_dict:
            output = (logits,) + outputs[2:]
            return ((loss,) + output) if loss is not None else output

        return TokenClassifierOutput(
            loss=loss,
            logits=logits,
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,
        )
```

CRF(Conditional Random Field)를 이용할 경우 `pytorch-crf` 모듈을 설치 후
```bash
pip install pytorch-crf
```

```python
from transformers import T5EncoderModel
from transformers.modeling_outputs import TokenClassifierOutput

import torch
import torch.nn as nn
from torchcrf import CRF

class T5EncoderForEntityRecognitionWithCRF(T5EncoderModel):
    def init(self, config):
        super().init(config)

        self.num_labels = config.num_labels

        self.dropout = nn.Dropout(config.dropout_rate)
        self.position_wise_ff = nn.Linear(config.d_model, config.num_labels)
        self.crf = CRF(config.num_labels, batch_first=True)

    def forward(
        self,
        input_ids=None,
        attention_mask=None,
        head_mask=None,
        inputs_embeds=None,
        labels=None,
        output_attentions=None,
        output_hidden_states=None,
        return_dict=None,
    ):
        return_dict = return_dict if return_dict is not None else self.config.use_return_dict

        if attention_mask is None:
            attention_mask = torch.ones(input_ids, device=input_ids.device)

        outputs = self.encoder(
            input_ids=input_ids,
            attention_mask=attention_mask,
            inputs_embeds=inputs_embeds,
            head_mask=head_mask,
            output_attentions=output_attentions,
            output_hidden_states=output_hidden_states,
            return_dict=return_dict,
        )

        last_hidden_state = outputs[0]
        last_hidden_state = self.dropout(last_hidden_state)
        emissions = self.position_wise_ff(last_hidden_state)

        loss = None
        if labels is not None:
            mask = attention_mask.to(torch.uint8)
            loss = self.crf(emissions, labels, mask=mask)
            loss = -1 * loss
            logits = self.crf.decode(emissions, mask)
        else:
            mask = attention_mask.to(torch.uint8)
            logits = self.crf.decode(emissions, mask)

        if not return_dict:
            output = (logits, ) + outputs[2:]
            return ((loss,) + output) if loss is not None else logits

        return TokenClassifierOutput(
            loss=loss,
            logits=logits,
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,
        )
```

로 모델을 만들어 사용하면 됩니다.

이렇게 만든 모델을 pre-trained weight를 이용하여 로드하려면 모델 생성 후 `from_pretrained` 메소드로 로드하면 됩니다.
푸시려고 하는 문제가 영역을 잡아내는 것이서 그친다면 BIO tagging을 한 class의 갯수는 3개 (["O", "B", "I"])가 될 것입니다.
만약 클래스가 N개라면 BIO 태깅한 클래스의 갯수는 2N+1이 될것입니다. ("태창위생지"-->"장소", "100"-->"양", 이 데이터 NER 같은데요... 아닌가요??)

클래수 갯수를 `num_labels`라고 할때, 다음과 같이 모델을 로드합니다.

```python
from transformers import AutoConfig

BIO_CLASS = ["O", "B", "I"]
num_labels = len(BIO_CLASS)
model_name_or_path = "KETI-AIR/ke-t5-base"
config = AutoConfig.from_pretrained(model_name_or_path , num_labels=num_labels)

model = T5EncoderForEntityRecognitionWithCRF.from_pretrained(model_name_or_path , config=config)
```

여기에 라벨과 tokenizer에서 나온 input_ids를 모델의 인풋으로 넣고 나오는 loss (첫번째 return arg)를 이용하여 학습하시면 될 것 같습니다.
단, 여기서 tokenize된 input ids와 BIO class를 맞춰주셔야 하는데요, fast tokenizer의 [char_to_token](https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.BatchEncoding.char_to_token)을 이용하시면 편리하실 겁니다.

**Comment by ahj6377** [comment]
답변 감사합니다!
제가 해결하고자 하는 task가 classification이 아니라 summarization 기반인데, short text에서부터
키워드를 추출하고자 하는거라 summarization 모델을 사용하고자 했습니다.
인코더에 sentence를 넣고, sentence에서부터 summarization 된 단어를 타겟값(=키워드)와의 차이를 비교해
학습시키고자 하는데, summarization 모델을 기반으로 학습을 하면 입력센텐스 -> T5 인코더 -> T5디코더 -> (디코딩한
키워드 <-> label값)을 비교하는 task를 사용하면 될까요?

2023년 5월 3일 (수) 오후 8:27, kimsan0622 ***@***.***>님이 작성:

**Comment by kimsan0622** [comment]
아하 그러시군요. 제 생각에는 keyword를 뽑아내기 위해서는 BIO 태깅을 이용한 classification이 효과적일거 같은데, 생성을 통해 하시려는 특별한 이유가 있으신가요??

뭐 어쨋든 만약 제가 생성을 통하여 keyword를 추출하려고 한다면, 입력 센턴스 -> T5 인코더-디코더 -> 키워드가 되도록 학습할 것 같습니다.
또한 이 경우 generation되는 token의 길이가 작기 때문에 generation 메소드의 max_length를 작게(16 정도??) 줄 것 같습니다. (num_beams: 4 등등 세팅해서)

그런데, 한 문장에서 keyword를 여러 개 뽑아내야하므로, 입력으로 입력문장 외에 몇번째 키워드를 뽑는지 알려주는 숫자를 덧붙일거 같습니다.
예를들어

| 입력 | 출력|
| --- | --- |
| "index: 0 sentence: 태창위생지의 행정동 이름이 뭐야" | "태창위생지" |
| "index: 1 sentence: 태창위생지의 행정동 이름이 뭐야" | "행정동 이름" |

이런식으로요.

더 이상 궁금한 점이 없으실 경우 이슈 클로즈 부탁드립니다.

**Comment by ahj6377** [comment]
생성을 통해 접근하려는 이유가 1. 먼저 레이블의 키워드 단어가 각자 제각각이라 레이블이 무한히 늘어날거 같다는점과
2. KeyphraseTransformer 논문을 읽어보니 그 논문에서는 summary -> generation을 통해 접근했기에
supervised learning으로 short sentence -> keyword generation을 통해 접근해보고자 했습니다!
답변 감사드립니다, 작성자님께서 남기신 답변 보고 잘 고민해보겠습니다!

2023년 5월 3일 (수) 오후 9:14, kimsan0622 ***@***.***>님이 작성:
